Save some memory by using memset and inlined memcpy

gudnimg opened the pull request:
The 27 commits save 1544 bytes of flash and 24 bytes of SRAM.

gudnimg committed:
Use memset to zero previous_speed array

Saves 150 bytes of flash memory

gudnimg committed:
Use inlined function set_destination_to_current() in more places

Saves 70 bytes of flash memory

gudnimg committed:
Replace for-loop with memcpy

Saves 22 bytes of flash memory

gudnimg commented:
Changes work fine on my end. 

![image](https://user-images.githubusercontent.com/8218499/129479733-e30b6bac-abab-49c0-b971-8cac2dbe4a85.jpeg)

![image](https://user-images.githubusercontent.com/8218499/129479748-800974a3-56ec-40a3-b33e-dbb79c722b05.jpeg)

gudnimg committed:
Change row_offsets[] to uint8_t

Saves 10 bytes of flash, and 4 bytes of SRAM

gudnimg committed:
Change return types of two functions uint8_t

These functions aren't used but they should return the appropriate type.

gudnimg committed:
Replace three lines by one function call which does the same thing

Saves 42 bytes of flash memory

gudnimg committed:
setTargetHotend expect an uint8_t, not int for second parameter

also change _usb_timer to ShortTimer

Saves 28 bytes of Flash and 1 byte of SRAM

gudnimg committed:
Convert mmu_last_request and mmu_last_response to ShortTimer

Saves 180 bytes of flash and 2 byte of SRAM

gudnimg committed:
Convert previous_millis_cmd to LongTimer

Saves 196 bytes of Flash but adds 1 byte of SRAM

gudnimg commented:
Wow I'm noticing there is a lot of room for optimisation by converting variables that use `_millis()` to `ShortTimer` or `LongTimer`

leptun commented:
Yeah. That's what I noticed as well. A lot of flash can be saved this way and sometimes even 1B of ram if ShortTimer is used. Otherwise, the ram usage increases by 1, but that can be easily recovered somewhere else.

gudnimg committed:
Convert autostart_atmillis to ShortTimer

Saves 34 bytes of flash and 1 byte of SRAM

gudnimg committed:
Convert NcTime to ShortTimer

The variable is only used in ultralcd.cpp so I made it static there
and deleted it from Marlin_main.cpp

Saves 24 bytes of flash and 1 byte of SRAM

gudnimg committed:
Convert mmu_last_finda_response to ShortTimer

Saves 48 bytes of Flash, 1 byte of SRAM

gudnimg committed:
Convert display_time to ShortTimer

Saves 24 bytes of flash, 1 byte of SRAM

gudnimg committed:
crashDetTimer can be ShortTimer

Expired value is less than 65535

Saves 8 bytes of flash and 2 bytes of SRAM

gudnimg commented:
Do not merge this yet, I will test the changes by printing stuff over the weekend ;D

gudnimg commented:
I believe we can use Timer.cpp/Timer.h in more places if we add the functionality to be able to read the variable `m_started` which stores the time when the timer first started `start()`.

I am specifically thinking about instances where the value is used in calculations. Divisions, subtraction between two timers etc.

leptun commented:
Yes. We can improve Timer.cpp/h. I was thinking also about a way of setting if it’s started by default (at construction)

gudnimg committed:
Remove unused kicktime timer

The variable is only set and never actually used

Saves 8 bytes of flash and 4  bytes of SRAM

leptun commented on Firmware/Marlin_main.cpp:
```diff
@@ -4652,8 +4650,6 @@ void process_commands()
       eeprom_update_word(reinterpret_cast<uint16_t *>(&(EEPROM_Sheets_base->s[(eeprom_read_byte(&(EEPROM_Sheets_base->active_sheet)))].z_offset)),0);
 
 	} else if(code_seen_P(PSTR("Beat"))) { // PRUSA Beat
```
Beat could be removed entirely if it is really unused.

gudnimg replied:
Yeah I'm not sure what it is supposed to do.

```
Set of internal PRUSA commands
    #### Usage
         PRUSA [ Ping | PRN | FAN | fn | thx | uvlo | MMURES | RESET | fv | M28 | SN | Fir | Rev | Lang | Lz | Beat | FR ]
```

It is an internal  Prusa command. Do you have any insight into it?

leptun replied:
@gudnimg Yeah. It should be “PRUSA Beat”. No idea if it is actually used. Since it does nothing, I think it can be removed.
@DRracer what do you think?

gudnimg replied:
I added a commit which removes it, if it is needed we can simply drop the commit to revert.

leptun replied:
Just curious, how does the firmware respond now to a “PRUSA Beat”?

gudnimg replied:
After I removed it, the printer replied with `"ok"` after I write `"PRUSA Beat"`

leptun replied:
Good. So no change in behaviour.

gudnimg commented:
By doing this we should be able to convert these variables:

- [ ] `starttime`
- [ ] `stoptime`
- [ ] `pause_time`
- [ ] `lcd_next_update_millis` (probably)
- [ ] `PingTime`
- [ ] `start_pause_print`
- [ ] `extruder_autofan_last_check`
- [ ] `temp_millis`, `t1`, `t2`

gudnimg committed:
Remove internal Prusa command 'Beat'

It doesn't seem to do anything.

Saves 22 bytes of flash

gudnimg committed:
Change usb_printing_counter to uint8_t

It is assigned max value of 10

Saves 26 byte of flash and 1 byte of SRAM

gudnimg committed:
Change heating_status to uint8_t

Saves 66 byte of flash and 1 byte of SRAM

gudnimg committed:
Change heating_status_counter to uint8_t

It has a max value of 14

Saves 28 bytes of flash and 1 byte of SRAM

gudnimg commented:
I think it would be a good idea to create an `uint8_t` enum for `heating_status`. Here is my inital proposal:

```c
enum class HeatingStatus : uint8_t
{
    NO_HEATING = 0,
    EXTRUDER_HEATING = 1,
    EXTRUDER_HEATING_COMPLETE = 2,
    BED_HEATING = 3,
    BED_HEATING_COMPLETE = 4
};
```

I wrote an implementation for this and I don't see any change in memory. Though the code may look more confusing.

gudnimg committed:
Change custom_message_state_old  and custom_message_state to uint8_t

I see max value as 7*7+10 = 59

Saves 100 byte of flash and 1 byte of SRAM

gudnimg committed:
Change farm_timer to uint8_t

Remove extern declaration of farm_timer and farm_status as they
are only used in one file.

Saves 26 bytes of flash

leptun commented:
I think it’s a lot better if enum class was used instead of whatever was before.

gudnimg committed:
Create enum class for heating_status variable

DRracer commented:
1KB of flash down is always welcome, we'll get to this in FW 3.11 roughly in late autumn 2021

gudnimg committed:
Remove variable move_menu_scale

The variable is just set to 1.0 and is never modified, I also
removed the function lcd_move_menu_1mm()

Changes save 96 bytes of flash memory and 4 bytes of SRAM

gudnimg committed:
Smaller code in G28

Replace two blocks of code with already defined functions.

Change saves 100 bytes of flash memory and 11 lines of code.

gudnimg committed:
G81: Change for loop variables x and y to uint8_t

Saves 198 bytes of flash, that's crazy

gudnimg committed:
Use code_value_short()  in get_command()

Same code but uses a defined function

Saves 28 bytes of flash memory

gudnimg committed:
Use code_value_uint8() in get_command()

The if statement is simplified as the value is never negative
and we can get rid of one variable

Saves 12 bytes of flash memory

leptun commented:
Bump so we don't forget about this
@DRracer @3d-gussner

DRracer commented:
@leptun yes, please make a PR so that I can tag it with milestone 3.11.0

leptun commented:
https://github.com/prusa3d/Prusa-Firmware/pull/3181

DRracer commented:
PR#3181 is already tagged with 3.11, so we won't forget it
